File: MediaBase/MediaBase/StreamInfo.m
//
//  StreamInfo.m
//  MediaBase
//
//  Created by tbago on 16/12/16.
//  Copyright © 2016年 tbago. All rights reserved.
//

#import "StreamInfo.h"

@interface StreamInfo()
@property (strong, nonatomic) NSMutableDictionary   *innerMetaData;
@end

@implementation StreamInfo

- (instancetype)init {
    self = [super init];
    if (self) {
        _duration = -1;
    }
    return self;
}

#pragma mark - public method

- (void)addMetaDataToStreamInfo:(NSString *)key
                          value:(NSString *)value {
    self.innerMetaData[key] = value;
}

#pragma mark - get & set

- (NSDictionary *)metaData {
    return _innerMetaData;
}

- (NSMutableDictionary *)innerMetaData {
    if (_innerMetaData == nil) {
        _innerMetaData = [[NSMutableDictionary alloc] init];
    }
    return _innerMetaData;
}

@end

File: MediaBase/MediaBase/StreamInfo.cc
//
//  StreamInfo.m
//  MediaBase
//
//  Created by tbago on 16/12/16.
//  Copyright © 2021 tbago. All rights reserved.
//

#include "StreamInfo.h"

namespace media_base {

StreamInfo::StreamInfo()
{
    streamType      = UnknownStream;
    streamId        = -1;
    bitrate         = -1;
    duration        = -1;

    width                     = -1;
    height                    = -1;
    framerate.numerator         = 1;
    framerate.denominator       = 1;
    pixelAspect.numerator       = 1;
    pixelAspect.denominator     = 1;

    sampleFormat                = R_SAMPLE_FMT_NB;
    samplerate                  = -1;
    channels                    = -1;

    extraData                   = NULL;
    extraDataSize               = 0;
}

StreamInfo::StreamInfo(const StreamInfo &streamInfo)
{
    CopyStreamInfo(streamInfo);
}

StreamInfo & StreamInfo::operator=(const StreamInfo &streamInfo)
{
    if (this == &streamInfo) {
        return *this;
    }
    CopyStreamInfo(streamInfo);
    return *this;
}

StreamInfo::~StreamInfo() {
}

void StreamInfo::CopyStreamInfo(const StreamInfo &streamInfo) {
    streamType = streamInfo.streamType;
    streamId = streamInfo.streamId;
    streamName = streamInfo.streamName;
    codecID = streamInfo.codecID;
    codecName = streamInfo.codecName;
    language = streamInfo.language;
    bitrate = streamInfo.bitrate;
    duration = streamInfo.duration;
    codecTag = streamInfo.codecTag;
    bitsPerCodedSample = streamInfo.bitsPerCodedSample;

    width = streamInfo.width;
    height = streamInfo.height;
    pixelFormat = streamInfo.pixelFormat;
    framerate = streamInfo.framerate;
    pixelAspect = streamInfo.pixelAspect;

    sampleFormat = streamInfo.sampleFormat;
    samplerate = streamInfo.samplerate;
    channels = streamInfo.channels;

    metaData.clear();
    for (auto it = streamInfo.metaData.begin(); it != streamInfo.metaData.end(); it++) {
        metaData[(*it).first] = metaData[(*it).second];
    }

    if (extraDataSize > 0 && extraData != NULL) {
        free(extraData);
    }

    if (streamInfo.extraDataSize > 0 && streamInfo.extraData != NULL) {
        extraDataSize = streamInfo.extraDataSize;
        extraData = (uint8_t *)malloc(extraDataSize);
        memcpy(extraData, streamInfo.extraData, extraDataSize);
    }
}
}   // media_base
